libraries/ms-common/src/stm32f0xx/spi.c

```diff
@@ -0,0 +1,185 @@
+  }
+
+  // Enable peripheral clock
+  if (spi_x == 1) {
+    RCC_APB2PeriphClockCmd(RCC_APB2Periph_SPI1, ENABLE);
+  } else {
+    RCC_APB1PeriphClockCmd(RCC_APB1Periph_SPI2, ENABLE);
+  }
+
+  // Note: gpio_init_pin initializes pin clocks as well
+
+  // Initialize MOSI pin
+  GPIOSettings pin_settings = { 0 };
+  pin_settings.direction = GPIO_DIR_OUT;
+  pin_settings.state = GPIO_STATE_LOW;
+  pin_settings.resistor = GPIO_RES_NONE;
+  pin_settings.alt_function = GPIO_ALTFN_0;
+  gpio_init_pin(&settings->mosi_pin, &pin_settings);
+
+  // Initialize SCK pin
+  gpio_init_pin(&settings->sck_pin, &pin_settings);
+
+  // Initialize MISO pin
+  pin_settings.direction = GPIO_DIR_IN;
+  pin_settings.resistor = GPIO_RES_PULLUP;
+  gpio_init_pin(&settings->miso_pin, &pin_settings);
+
+  // Initialize CS pin
+  pin_settings.direction = GPIO_DIR_OUT;
+  pin_settings.state = GPIO_STATE_HIGH;
+  gpio_init_pin(&settings->cs_pin, &pin_settings);
+  s_cs_pin_map[spi_x] = settings->cs_pin;
+
+  // Fill SPI init struct
+  SPI_InitTypeDef spi_init_struct;
+
+  // Default settings
+  spi_init_struct.SPI_Direction = SPI_Direction_2Lines_FullDuplex;
+  spi_init_struct.SPI_Mode = SPI_Mode_Master;
+  spi_init_struct.SPI_NSS = SPI_NSS_Soft;
+  spi_init_struct.SPI_CRCPolynomial = 7;
+  spi_init_struct.SPI_DataSize = SPI_DataSize_8b;
+
+  // Parse clock polarity
+  if (settings->polarity == SPI_CPOL_LOW) {
+    spi_init_struct.SPI_CPOL = SPI_CPOL_Low;
+  } else {
+    spi_init_struct.SPI_CPOL = SPI_CPOL_High;
+  }
+
+  // Parse clock phase
+  if (settings->phase == SPI_CPHA_1EDGE) {
+    spi_init_struct.SPI_CPHA = SPI_CPHA_1Edge;
+  } else {
+    spi_init_struct.SPI_CPHA = SPI_CPHA_2Edge;
+  }
+
+  // Parse Baud Rate prescaler value
+  spi_init_struct.SPI_BaudRatePrescaler = ((settings->baud_rate / 2) << 4)
+                                      + ((settings->baud_rate % 2) << 3);
+
+  // Parse first bit
+  if (settings->first_bit == SPI_FIRSTBIT_MSB) {
+    spi_init_struct.SPI_FirstBit = SPI_FirstBit_MSB;
+  } else {
+    spi_init_struct.SPI_FirstBit = SPI_FirstBit_LSB;
+  }
+
+  Spi_Init(spi_periph_map[spi_x], &spi_init_struct);
+
+  // Enable NVIC for interrupts
+  NVIC_InitTypeDef nvic_init_struct = { 0 };
+  nvic_init_struct.NVIC_IRQChannel = SPI1_IRQn + spi_x;
+  nvic_init_struct.NVIC_IRQChannelPriority = 1; // no idea if this is correct priority
+  nvic_init_struct.NVIC_IRQChannelCmd = ENABLE;
+
+  // Set FIFO threshold for RXNE event
+  SPI_RxFIFOThresholdConfig(spi_periph_map[spi_x], SPI_RxFIFOThreshold_QF);
+
+  // Enable peripheral
+  SPI_Cmd(spi_periph_map[spi_x], ENABLE);
+  return STATUS_CODE_OK;
+}
+
+StatusCode spi_exchange(uint8_t spi_x, uint8_t *tbuf, uint8_t *rbuf,
+  size_t t_length, size_t r_length) {
+  if (!prv_is_periph_valid(spi_x)) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  s_spi_x = spi_x;
+
+  for (int i = 0; i * sizeof(uint8_t) < (t_length > r_length ? t_length : r_length); ++i) {
```

Comment:
I'm not entirely sure what's going on in this line. Multiplying by sizeof(uint8_t) does nothing because sizeof(uint8_t) == 1, and you wouldn't want to be multiplying `i` by it even if it wasn't, since it's an array index and you aren't doing direct pointer math. (fun fact: C has pointer arithmetic, which means `uint64_t *pointer = 0x1234; pointer++;` will actually increment `pointer` by sizeof(uint64_t), not 1.)

Also, SPI exchanges will usually have an array of data sent from the master to the slave where the slave's response is all don't cares (usually 0x00). Then, the master sends 0x00 to get the slave's actual response. So you just need two loops, one where the master is sending the tx buffer, and the other where they're sending 0x00 and reading into the rx buffer.